### ISSUE TITLE
(Discussion) Static Site Generation in DSpace

### ISSUE THREAD

**Issue by kshepherd** [DSpace/dspace-angular#3183]
## Description

Generating and serving static HTML and javascript ('Static Site Generation') can help make repository content easier to run in low resource environments, more portable (pages are on disk), and checks off many other best practices.

DSpace Angular can't do everything as static content, of course, but there are many pages which don't change very much (e.g. unauthenticated context + item page) and are the most commonly visited content, by both human and robot users.

This ticket is for discussion about how and where we can fit SSG (static site generation) into DSpace, techniques and how they work, and any experience or progress experimenting with this topic.

## Approaches

There seem to be a few approaches with SSG in Angular, so far we have identified:

1. Build-time generation of configured or discovered routes (user-specified) - this could hammer the REST API quite hard at startup. It could be more useful for repositories who have large numbers of item pages that do not change for years.
2. On-demand / 'regeneration' as requests come in? _need to expand this description_
3. External tools like scully (see Refs) that can crawl the Angular site and create static content, could be configured and used to generate things incrementally or as admins see fit, on TTLs, etc.

### Note re: performance

This discussion is somewhat related to performance but is not intended to be a band-aid "fix" for high-resource usage in SSR (server-side rendering) - it may mitigate some issues but is a valuable goal in an of itself.

## References

* https://v17.angular.io/guide/prerendering documents some build-time SSG
* https://github.com/scullyio/scully is an external tool that can guess routes / js and generate static content for some Angular sites
* https://blog.angular.dev/incremental-static-regeneration-for-angular-c12745ff4b48 documents some incremental static regeneration techniques
* https://zenodo.org/records/8111569 is the OR2023 closing keynote which inspired the OR2024 static site workshop which @artlowel and @kshepherd have referenced a few times in dev meetings

**Comment by abollini** [comment]
please please please :) only look for **option 3**.
IMHO we need to simplify as much as possible the angular code to reach our primary goal that should be have a fast application in a basic scenario without the need of all these extra layers (cache, pre-site generation) that are of course needed for High Performance, large and heavy accessed site.
The other options could be attractive in the short term but would make the performance worst, the installation process more complex and/or slow

**Comment by kshepherd** [comment]
> please please please :) only look for **option 3**

Good points made today, I've added notes about the downsides to options 1 and 2, and tried to make it clearer that this issue was not supposed to be related to SSR (or even Angular, necessarily) or compete for attention with SSR performance issues.

**Comment by jameswsullivan** [comment]
I don't know much about the inner workings of DSpace but I really wish that there could be something like the "Simply Static" WordPress plugin for it.

**Comment by kshepherd** [comment]
@jameswsullivan would a function like "generate website from publicly-accessible resources" fit your use case? (i.e. any items, bitstreams, collections that are not readable by unauthenticated users would be excluded)

**Comment by jameswsullivan** [comment]
@kshepherd I think so, yes.
One of the performance issues we have is that the DSpace site is getting hit by large amounts of either crawling or content scraping bots that enumerate the links/resources, causing heavy load to the hosting. And I think under DSpace's current structure, each of these sessions would trigger calls/responses between the DSpace angular UI and the DSpace API, and the SSR stuff? (I have limited knowledge about how exactly DSpace works but these are the topics I've come across during hosting and orchestration.)

So I'd think if publicly-accessible resources could be generated and then hosted/served in a static way (especially for the bots and anonymous hits), that would alleviate a lot of this problem? But I'm not sure how it would work though, will the publicly-accessible content be served as static pages first, until a user logs in? My analogy to WordPress' Simply Static is probably not a good fit in this scenario because I simply spin up a local WordPress instance and make edits, and then generate the static HTMLs and upload them to the hosting server.

**Comment by saschaszott** [comment]
@kshepherd , I think the SSG approach could work well, as long as it is properly integrated into the overall DSpace backend architecture.

I ran a few tests to generate static HTML versions of item landing pages using **Playwright** ([https://playwright.dev/docs/intro](https://playwright.dev/docs/intro)). Playwright renders the dynamic page content and retrieves the metadata from the backend through the REST API. The rendered HTML was then saved locally in files (one HTML file per landing page).

For nearly 8000 publication items in my test env, this approach produced almost 5 GB of HTML content. The main issue is that Angular-generated HTML is **very verbose**. In my tests, a single landing page was around 0.5 - 0.6 MB, even though the actual content data is only a few kilobytes.

For that reason, I do not think rendering the existing Angular pages and storing the result is practical. Since landing pages, especially for publication items, rarely change after the item is archived, IMHO it would make more sense to generate static HTML directly in the backend without Angular. This HTML should be kept deliberately lean, with as little additional markup and structure as possible.

**Comment by saschaszott** [comment]
One could also consider limiting the number of search and browsing options available to unauthenticated users, which would include bots. Bots would not lose access to any archived item stored in the repository, since they could systematically retrieve all item landing pages through the sitemap and would not need search or browse functionality. Human users could register if they require additional search or browsing features.

**Comment by kshepherd** [comment]
@saschaszott I agree with you, serving entire webpacks / one-page apps as a "page" defeats the purpose, and this is why I ended up going down different routes as linked in this issue description.

This is a slight tangent, but I personally have a half-baked idea of a future DSpace architecture where we have "Backoffice" frontend+backend modules (angular + REST + DB +Solr or some kind of fully featured web app) for repository staff to use, a "Submission / Workspace / Workflow" module for submitters and other non-staff users, made somewhat more accessible but still not available to harvesters, and then a very lightweight static website with search/browse and other simple functionality which is easy to render into lean HTML + Solr/ES search.
Only that last module would need to be published to anonymous Internet users and it would all be serving plain files from disk with some lightweight search indexing/faceting/filtering supplied by Solr or ES.
We have protocols like SWORD to help separate submission from the rest of the admin features, and SWORDv2 ([v3 also exists](https://swordapp.github.io/swordv3/swordv3.html)) understands the idea of workspace / different deposit states.

The main problem that I haven't easily solved (in my head) yet is authorization in the static/public website, so right now the idea mainly suits  a fully OA repository website. Depending on complexity, it might be possible to use htaccess type approaches for authZ with authN handled in a gateway/frontend layer.
